src/Storages/StorageMongoDB.cpp

```diff
@@ -87,6 +145,72 @@ void StorageMongoDB::connectIfNotConnected()
     }
 }
 
+ColumnsDescription StorageMongoDB::getTableStructureFromData()
+{
+    connectIfNotConnected();
+
+    MongoDBCursor cursor(database_name, collection_name, {}, {}, *connection, false);
+    FormatSettings format_settings;
+
+    std::unordered_map<std::string, DataTypePtr> types;
+    std::vector<std::string> current_keys;
+    std::vector<std::string> final_keys;
+
+    size_t num_rows = 0;
+    while (num_rows < format_settings.max_rows_to_read_for_schema_inference)
+    {
+        auto documents = cursor.nextDocuments(*connection);
+
+        for (auto & document : documents)
+        {
+            if (document->exists("ok") && document->exists("$err")
+                && document->exists("code") && document->getInteger("ok") == 0)
+            {
+                auto code = document->getInteger("code");
+                const Poco::MongoDB::Element::Ptr value = document->get("$err");
+                auto message = static_cast<const Poco::MongoDB::ConcreteElement<String> &>(*value).value();
+                throw Exception(ErrorCodes::MONGODB_ERROR, "Got error from MongoDB: {}, code: {}", message, code);
+            }
+
+            ++num_rows;
+
+            document->elementNames(current_keys);
+
+            for (auto & key : current_keys)
+            {
+                const Poco::MongoDB::Element::Ptr value = document->get(key);
+                auto new_type = getDataTypeMongoDB(*value);
+
+                if (!types.contains(key))
+                {
+                    types[key] = std::move(new_type);
+                    final_keys.push_back(std::move(key));
+                    continue;
+                }
+
+                auto & type = types[key];
+                transformInferredTypesIfNeeded(type, new_type, format_settings);
```

Comment:
I am not sure we should use `transformInferredTypesIfNeeded` here, because it was intended to use for text formats when we can have Date/DateTime as String. I think we can simply use `getLeastSuperType` here